src/solvers/petsc_dm_wrapper.C

```diff
@@ -16,39 +16,355 @@
+            }
+
+          // Lastly, Compute the subsection for the subDM
+          ierr = DMCreateSubDM_Section_Private(dm, numFields, fields, is, subdm);
+          LIBMESH_CHKERR(ierr);
+        }
+
+      return 0;
+    }
+
+    //! Help PETSc identify the finer DM given a dmc
+    PetscErrorCode __libmesh_petsc_DMRefine(DM dmc, MPI_Comm comm, DM * dmf)
+    {
+      libmesh_assert(dmc);
+      libmesh_assert(dmf);
+      libmesh_assert(comm);
+
+      PetscErrorCode ierr;
+
+      // extract our context from the incoming dmc
+      void * ctx_c = NULL;
+      ierr = DMShellGetContext(dmc, & ctx_c);LIBMESH_CHKERR(ierr);
+      libmesh_assert(ctx_c);
+      PetscDMContext * p_ctx = static_cast<PetscDMContext * >(ctx_c);
+
+      // check / set the finer DM
+      libmesh_assert(p_ctx->finer_dm);
+      libmesh_assert(*(p_ctx->finer_dm));
+      *(dmf) = *(p_ctx->finer_dm);
+
+      return 0;
+    }
+
+    //! Help PETSc identify the coarser DM dmc given the fine DM dmf
+    PetscErrorCode __libmesh_petsc_DMCoarsen(DM dmf, MPI_Comm comm, DM * dmc)
+    {
+      libmesh_assert(dmc);
+      libmesh_assert(dmf);
+      libmesh_assert(comm);
+
+      PetscErrorCode ierr;
+
+      // Extract our context from the incoming dmf
+      void * ctx_f = NULL;
+      ierr = DMShellGetContext(dmf, &ctx_f);LIBMESH_CHKERR(ierr);
+      libmesh_assert(ctx_f);
+      PetscDMContext * p_ctx = static_cast<PetscDMContext*>(ctx_f);
+
+      // First, ensure that there exists a coarse DM that we want to
+      // set. There ought to be as we created it while walking the
+      // hierarchy.
+      libmesh_assert(p_ctx->coarser_dm);
+      libmesh_assert(*(p_ctx->coarser_dm));
+
+      // In situations using fieldsplit we need to (potentially)
+      // provide a coarser DM which only has the relevant subfields in
+      // it. Since we create global DMs for each mesh level, we need
+      // to extract the section from the DM, and check the number of
+      // fields. When less than all the fields are used, we need to
+      // create the proper subsections.
+
+      // Get the number of fields and their names from the incomming
+      // fine DM and the global reference DM
+      PetscInt nfieldsf, nfieldsg;
+      char ** fieldnamesf;
+      char ** fieldnamesg;
+
+      libmesh_assert(p_ctx->global_dm);
+      DM * globaldm = p_ctx->global_dm;
+      ierr = DMCreateFieldIS(dmf, &nfieldsf, &fieldnamesf, NULL);
+      LIBMESH_CHKERR(ierr);
+      ierr = DMCreateFieldIS(*globaldm, &nfieldsg, &fieldnamesg, NULL);
+      LIBMESH_CHKERR(ierr);
+
+      // If the probed number of fields is less than the number of
+      // global fields, this amounts to PETSc 'indicating' to us we
+      // are doing FS. So, we must create subsections for the coarser
+      // DMs.
+      if ( nfieldsf < nfieldsg )
+        {
+          PetscSection section;
+          PetscSection subsection;
+          PetscInt * subfields = new PetscInt[nfieldsf]; // extracted fields
```

Comment:
I'd prefer a `std::vector` here, and then just pass the `.data()` array to PETSc to use. Then you don't have to remember to `free []` it later.

Reply:
Ahh, I hated this part of the code and didn't know about this trick, thanks!